samusrgrp.c

```diff
@@ -217,11 +314,36 @@ int main(int argc, char **argv)
     exit(1);
   }
 
-  if (!list && (!usr || !grp || !*usr || !*grp) ) {
-    fprintf(stderr,"%s: ERROR: Both -u and -g must be specified.\n",argv[0]);
-    exit(1);
-  }
+  if ( list == 0 )
+  {
+      if ( !grp || !*grp )
+      {
+          fprintf(stderr,"%s: ERROR: -g must be specified with -a or -r.\n",argv[0]);
+          exit(1);
+      }
+
+      if ( ( sid && usr ) || ( !sid && !usr ) )
+      {
+          fprintf(stderr,"%s: ERROR: -u or -S must be specified with -a or -r.\n",argv[0]);
+          exit(1);
+      }
+
+      if ( sid && !*sid )
+      {
+          fprintf(stderr,"%s: ERROR: SID must be specified with -S.\n",argv[0]);
+          exit(1);
+      }
+
+      if ( usr && !*usr )
+      {
+          fprintf(stderr,"%s: ERROR: SID must be specified with -S.\n",argv[0]);
+          exit(1);
+      }
+
+      if ( sid ) use_sid = 1;
+      if ( usr ) use_sid = 0;
```

Comment:
There shouldn't be an additional line before the closing } in this function in order to match current style.